### ISSUE TITLE
Usage behind proxy

### ISSUE THREAD

**Issue by nilsga** [slackhq/node-slack-client#33]
I'm trying to use this library behind a corporate proxy. How can this be configured, if possible?

**Comment by evansolomon** [comment]
I have no specific experience but would guess it works like any other requests from behind the proxy. One difference may be that the client uses a web socket, not sure if that's the sort of thing that has to be explicitly whitelisted.

Are you having trouble with it?  Any specific error messages?  Can you connect to other web socket servers (e.g. https://www.websocket.org/echo.html)?

**Comment by nilsga** [comment]
I have been investigating this some more, and I have come to the following conclusion. To be able to use this library through a proxy, it needs to:

* Use the `http` module instead of `https` module, and connect to the API-server through a proxy, as described here: http://stackoverflow.com/questions/16459751/calling-a-https-web-service-in-node-js-behind-proxy
* Somehow achieve the same with the `ws` module, which also uses `https` module instead of `http` module when using the `wss://` protocol.

**Comment by evansolomon** [comment]
In the StackOverflow example, they're making a request to an entirely different host.  Instead of https://x.com, they're requesting http://proxy:8080/x.com.  So I assume anything is going to require some configuration by the user since the library obviously can't know some other address to use.

From a quick look at the `Client` class, it seems like two things might be a good start.

1. Changing the `host` property of a client instance
2. Allowing some way to prefix/customize the `socketUrl`

**Comment by nilsga** [comment]
The "entirely different host" is the proxy server, and the format of the request is apparently how a request is proxied through a http proxy at the http level. So the supporting http/https libraries used by this module, and the websocket module does not transparently support proxy configuration. That would have to be done by this module, or depend on modules that can handle it, I suppose.

**Comment by evansolomon** [comment]
> The "entirely different host" is the proxy server

Yea, I got that.  Sorry if my comment about it was unclear.

As someone who never uses this sort of proxy, I don't really know enough to say exactly what the best way to do that is (my 2 steps above were a very rough guess of where one might start).  A PR for this feature would be cool :)

**Comment by nilsga** [comment]
Sorry, didn't mean sound rude. I just wanted to clarify the usage of proxy through the node http/https module. I don't think it's a quick solution to this either. It's more of a general NodeJS issue I guess.

**Comment by jtatum** [comment]
The requests library has a bunch of code for proxy support and might be a good place to find inspiration. It's a little surprising the functionality isn't built into node. https://github.com/request/request

**Comment by evansolomon** [comment]
@nilsga you didn't sound rude at all!

**Comment by structAnkit** [comment]
@nilsga I have been having this issue for a few months and decided yesterday to finally roll up my sleeves and figure this out.

I found that modifying slack-client to take an `agent` parameter to use in its `https.request` options works, assuming you supply something like the `HttpProxyAgent` module as the agent. Unfortunately this only gets you past the initial `rtm.start` call; the `ws` module also needs the agent. I tried hacking around and found out that `ws` uses a module called `Options` which messes with any agent object you pass into the `WebSocket` initializer.

My final solution was to test things like `tunnel`, `global-tunnel`, `proxy-tunnel`, and `proxy-out`, all of which failed miserably for a variety of reasons (the first three haven't been updated in years to support Node 0.12.x, and `proxy-out` for some reason is yielding 503 API errors for me). So what I did was simply override the `https.request` method to add the `HttpsProxyAgent` as the agent before creating the request:

```js
var HttpsProxyAgent = require('https-proxy-agent')
var agent = new HttpsProxyAgent('http://127.0.0.1:3128')

var https = require('https')
var httpsRequest = https.request
https.request = function(options, callback) {
  // Set agent if agent not already supplied and if HttpsProxyAgent is instantiated
  if (!options.agent && agent) {
    options.agent = agent
  }
  return httpsRequest(options, callback)
}
```
